### ISSUE TITLE
Replace usage of t.p.compat.reduce in core

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#6565]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)| @thijstriemstra reported|
|-|-|
|Trac ID|trac#6565|
|Type|task|
|Created|2013-06-10 18:43:50Z|
|Branch|https://github.com/twisted/twisted/tree/core-reduce-6565|

The builtin `reduce()` was moved to [functools.reduce](http://docs.python.org/2/library/functools.html#functools.reduce) in python2.6 and up and the builtin was removed in Python 3.

We can replace the usage of `twisted.python.compat.reduce` (introduced in r27667 for compatibility with python < 2.6) and usage of the builtin `reduce` with `functools.reduce` instead and eventually deprecate it.

```
twisted/protocols/ftp.py:32:from twisted.python.compat import reduce
twisted/protocols/ftp.py:1796:        return reduce(filepath.FilePath.child, path, self.filesystemRoot)
twisted/protocols/policies.py:199:        self.factory.registerWritten(reduce(operator.add, map(len, seq)))
twisted/protocols/gps/rockwell.py:96:      cksum = -(reduce(operator.add, (sync, msg_id, length, acknak)) & 0xFFFF)
twisted/protocols/gps/rockwell.py:123:        cksum = 0x10000 - (reduce(operator.add, struct.unpack('<%dH' % (length/2), message)) & 0xFFFF)
twisted/protocols/gps/nmea.py:24:from twisted.python.compat import reduce
twisted/protocols/gps/nmea.py:75:            checksum, calculated_checksum = int(checksum, 16), reduce(operator.xor, map(ord, strmessage))
twisted/spread/jelly.py:104:from twisted.python.compat import reduce
twisted/spread/jelly.py:588:        value = reduce(lambda left, right: left * 10 + right, guts)
```

Attachments:

* [remove-twisted-compat-reduce.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/741/7419673903e9bfd912a49fdce8f409519602d30f/1d80e3bd5eaf6a83e3a7be760746117aacb23c4d.patch) (12069 bytes) - added by ihrachyshka on 2013-07-06 11:36:15Z - Removed compat reduce

<details><summary>Searchable metadata</summary>

```
trac-id__6565 6565
type__task task
reporter__thijs thijs
priority__low low
milestone__Python_3_x Python-3.x
branch__branches_core_reduce_6565 branches/core-reduce-6565
branch_author__thijs thijs
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1370889830000000 1370889830000000
changetime__1383594917000000 1383594917000000
version__None None
owner__thijs thijs
cc__thijs
```
</details>

**Comment by twisted-trac** [comment]
|<img alt="ihrachyshka's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|ihrachyshka set owner to ihrachyshka<br>ihrachyshka set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|<img alt="ihrachyshka's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|ihrachyshka removed owner<br>ihrachyshka set status to `new`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_4"></a>|@exarkun commented|
|-|-|

Thanks for your work on this.

  1. We cannot remove `twisted.python.compat.reduce`.  See the ticket description which alludes to the [deprecation process](https://twistedmatrix.com/trac/wiki/CompatibilityPolicy).
  1. Likewise, the unit tests for `twisted.python.compat.reduce` should remain.
  1. Please don't rewrite "import x, y" as "import x\nimport y" (nor the reverse).  Either of these styles is accepted by the Twisted coding standard.
  1. Some of the new lines use more than 80 columns, please limit lines to 80 columns or fewer
  1. In `twisted/protocols/ftp.py`, the use of `reduce` can be replaced by `FilePath.descendant`
  1. Using `from functools import reduce` will reduce the size of the diff in some places.
  1. The changes in `twisted/protocols/gps/rockwell.py` seem to be untested
  1. The changes in `twisted/protocols/policies.py` seem to be untested
  1. The changes in `twisted/words/protocols/msn.py` seem to be untested
  1. There's also a typo in the docstring of the changes function in `msn.py` that it would be nice to fix.
  1. The changes in `twisted/words/protocols/oscar.py` seem to be untested

Some of the test coverage shouldn't be very hard to add - I suggest at least adding it for `policies.py`, as that is relatively core, useful functionality.  You may want to consider splitting the other untested changes off into separate tickets, though (or perhaps one for rockwell and one for the stuff in twisted/words/).

Thanks again.

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_5"></a>|@thijstriemstra commented|
|-|-|

Replying to [exarkun](#note_4):
> You may want to consider splitting the other untested changes off into separate tickets, though (or perhaps one for rockwell and one for the stuff in twisted/words/).

Yea, there are separate tickets for words ([#6564](https://github.com/twisted/twisted/issues/6564)) and names ([#6563](https://github.com/twisted/twisted/issues/6563)).

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_6"></a>|@thijstriemstra commented|
|-|-|

(In [40076]) Branching to 'core-reduce-6565'

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_7"></a>|@thijstriemstra commented|
|-|-|

(In [40077]) apply relevant part of remove-twisted-compat-reduce.patch, refs [#6565](https://github.com/twisted/twisted/issues/6565)

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_8"></a>|@thijstriemstra commented|
|-|-|

(In [40081]) address review comments, refs [#6565](https://github.com/twisted/twisted/issues/6565)

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_9"></a>|@thijstriemstra commented|
|-|-|

Replying to [exarkun](#note_4):
>   1. Some of the new lines use more than 80 columns, please limit lines to 80 columns or fewer

Fixed.

>   1. In `twisted/protocols/ftp.py`, the use of `reduce` can be replaced by `FilePath.descendant`

Done.

>   1. Using `from functools import reduce` will reduce the size of the diff in some places.

Done.

>   1. The changes in `twisted/protocols/gps/rockwell.py` seem to be untested

The only change is a that functools.reduce gets imported directly instead of from twisted.python.compat, so does this need a test? I see there are no tests at all and it seems unmaintained..

>   1. The changes in `twisted/protocols/policies.py` seem to be untested

Added a test.


Forced a [build](http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/core-reduce-6565).

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_10"></a>|@itamarst set owner to @thijstriemstra|
|-|-|

1. In the new test, please replace "port" variable name with the more accurate "protocol".
2. File a ticket for writing tests (or perhaps for deprecating) the untested GPS code if one does not already exist.

Otherwise looks good, please merge.

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_11"></a>|@thijstriemstra commented|
|-|-|

(In [40574]) address review comments, refs [#6565](https://github.com/twisted/twisted/issues/6565)

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_12"></a>|@thijstriemstra set status to `assigned`|
|-|-|

Thanks for the review.

Replying to [itamar](#note_10):
> 1. In the new test, please replace "port" variable name with the more accurate "protocol".

Fixed.

> 2. File a ticket for writing tests (or perhaps for deprecating) the untested GPS code if one does not already exist.
> 

Opened [#6810](https://github.com/twisted/twisted/issues/6810).

> Otherwise looks good, please merge.

Forced a [build](http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/core-reduce-6565), will merge when it's green.

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_13"></a>|@thijstriemstra set status to `closed`|
|-|-|

(In [40575]) Merge core-reduce-6565: Replace usage of twisted.python.compat.reduce in core.

Author: thijs, ihrachyshka
Reviewer: exarkun, itamar
Fixes: [#6565](https://github.com/twisted/twisted/issues/6565)

The builtin `reduce` was moved to `functools.reduce` and deprecated in py2.6 (and removed in python3). Replacing the usage of twisted.python.compat.reduce with functools.reduce in twisted core so compat.reduce can be deprecated eventually.
